test/integration/test_timeout.py

```diff
@@ -0,0 +1,214 @@
+#  © Copyright IBM Corporation 2020.
+#  #
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#  #
+#       http://www.apache.org/licenses/LICENSE-2.0
+#  #
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import time
+import unittest
+
+import requests
+from unittest.mock import Mock
+from requests import Response
+from urllib3 import Timeout
+
+from ibmcloudant.cloudant_v1 import CloudantV1
+from ibm_cloud_sdk_core.authenticators import BasicAuthenticator, IAMAuthenticator, NoAuthAuthenticator
+from ibmcloudant.couchdb_session_authenticator import CouchDbSessionAuthenticator
+
+DEFAULT_TIMEOUT = 150 # 2.5m (=150s)
+CUSTOM_TIMEOUT = 10   # 10s
+
+class TestTimeout(unittest.TestCase):
+
+# ** Helpers
+# *************
+
+    _timeout_mock_response = Response()
+    _timeout_mock_response.status_code=200
+    _timeout_mock_response._content = '{"foo":"bar"}'.encode()
+    _timeout_mock_response.cookies.set("AuthSession", "bar", expires=time.time()+120)
+    _custom_timeout={'timeout':CUSTOM_TIMEOUT}
+
+    def _set_no_timeout(self, new_value:dict):
+        new_value['timeout']
+
+    # Every test case tests an authenticator and its default timeout value at first,
+    # in the second iteration a custom overwrite possibility is checked.
+    def _defineTestCases(self):
+        return [
+            {
+                'assert_func': self._assert_default_timeout_setting,
+                'set_timeout': self._set_no_timeout
+            },
+            {
+                'assert_func': self._assert_custom_timeout_setting,
+                'set_timeout': self.cloudant.set_http_config
+            }
+        ]
+
+    def _mock_out_cloudant_request(self, _mock_response=_timeout_mock_response):
+        self.cloudant.http_client.request = Mock(return_value=_mock_response)
+
+    def _get_authenticate_arguments(self):
+        requests.request.assert_called_once()
+        return requests.request.mock_calls[0]
+
+    def _get_request_arguments(self, counter):
+        self.assertEqual(self.cloudant.http_client.request.call_count, counter+1)
+        return self.cloudant.http_client.request.mock_calls[counter]
+
+    def _assert_default_timeout_setting(self, call_args):
+        _, _, kwargs = call_args
+        self.assertTrue(isinstance(kwargs['timeout'], Timeout))
+        self.assertEqual(kwargs['timeout'].read_timeout, DEFAULT_TIMEOUT)
+
+    def _assert_custom_timeout_setting(self, call_args):
+        _, _, kwargs = call_args
+        self.assertEqual(kwargs['timeout'], CUSTOM_TIMEOUT)
+
+
+# ** CloudantV1
```

Comment:
We may want a `tearDown` step to clear the variables after each test e.g.
`del self.cloudant.http_client.request`

Reply:
I'm wondering if we should use `setUp` to initialize the mock variables e.g.
```
    def setUp(self):
      self.timeout_mock_response = Response()
      self.timeout_mock_response.status_code=200
      self.timeout_mock_response._content = '{"foo":"bar"}'.encode()
      self.timeout_mock_response.cookies.set("AuthSession", "bar", expires=time.time()+120)
      self.custom_timeout={'timeout':CUSTOM_TIMEOUT}
```

Reply:
> We may want a `tearDown` step to clear the variables after each test e.g.
`del self.cloudant.http_client.request`

I am not get it, why would such deletion would be needed? `self.cloudant` will be reinitialized in every test case. I can convert `cloudant` variable to local if you think that would be more secure.

> I'm wondering if we should use `setUp` to initialize the mock variables e.g.
```
    def setUp(self):
      self.timeout_mock_response = Response()
      self.timeout_mock_response.status_code=200
      self.timeout_mock_response._content = '{"foo":"bar"}'.encode()
      self.timeout_mock_response.cookies.set("AuthSession", "bar", expires=time.time()+120)
      self.custom_timeout={'timeout':CUSTOM_TIMEOUT}
```

I think the `setUp` runs before each test case, I think there is no need for this re initialization here. I wanted to define class level variables, that could be used by every method, I could also call them class level constants.

Reply:
Changed in 7c0ccbf

Reply:
I now see that the previous version was a mess from this perspective.

I have also found this:
> The testing code of a TestCase instance should be entirely self contained, such that it can be run either in isolation or in arbitrary combination with any number of other test cases.

Now I followed this idiom here: 21c626a